File: weave_introduction_handson/en/2_1_assets.py
"""
2_1: Assets and Scorers - Asset management and evaluation functions

What you'll learn in this script:
================================
1. weave.Model - Version LLM wrappers
2. weave.StringPrompt / weave.MessagesPrompt - Version prompts
3. weave.Dataset - Create and retrieve datasets
4. Scorer - Create and publish evaluation functions
5. Built-in Scorers - Use scorers provided by Weave

Where to look after running:
================================
- Models / Prompts / Datasets: Published assets
- Scorers: Evaluation functions you created
"""

import asyncio
import json
import os
from pathlib import Path

from dotenv import load_dotenv
from openai import OpenAI
import weave
from weave import Dataset, Scorer

from config_loader import chat_completion, get_model_name

# Load environment variables
load_dotenv()

# Initialize Weave
# Initialize with weave.init("entity/project")
ENTITY = os.getenv("WANDB_ENTITY")
PROJECT = os.getenv("WANDB_PROJECT", "weave-handson")
weave.init(f"{ENTITY}/{PROJECT}")

# Path to store asset references
ASSETS_FILE = Path(__file__).parent.parent / "assets.json"


def save_asset_ref(asset_type: str, name: str, ref: str):
    """Save an asset reference to a local file."""
    assets = json.load(open(ASSETS_FILE)) if ASSETS_FILE.exists() else {}
    assets.setdefault(asset_type, {})[name] = ref
    json.dump(assets, open(ASSETS_FILE, "w"), indent=2)
    print(f"Saved {asset_type}/{name}: {ref}")


# =============================================================================
# 1. weave.Model - Version an LLM wrapper
# =============================================================================
print("\n" + "=" * 60)
print("1. weave.Model - Version an LLM wrapper")
print("=" * 60)


class GrammarCorrector(weave.Model):
    """Grammar correction model.

    By inheriting from weave.Model and decorating predict with @weave.op(),
    configuration such as system_message / model_name / temperature and
    inference code can be versioned together.
    """

    system_message: str
    model_name: str
    temperature: float = 0.0

    @weave.op()
    def predict(self, sentence: str) -> dict:
        oai = OpenAI()
        response = oai.chat.completions.create(
            model=self.model_name,
            messages=[
                {"role": "system", "content": self.system_message},
                {"role": "user", "content": sentence},
            ],
            temperature=self.temperature,
        )
        return {"corrected": response.choices[0].message.content}


model_v1 = GrammarCorrector(
    system_message="You are a grammar checker. Return only the corrected sentence.",
    model_name=get_model_name(),
)
result = model_v1.predict("She goed to the store yesterday.")
print(f"v1 result: {result}")

model_v2 = GrammarCorrector(
    system_message="You are an expert grammar checker. Return only the corrected sentence, no explanation.",
    model_name=get_model_name(),
)
result = model_v2.predict("I has a big dog.")
print(f"v2 result: {result}")

model_ref = weave.publish(model_v2, name="grammar_corrector")
save_asset_ref("models", "grammar_corrector", model_ref.uri())

retrieved_model = weave.ref(model_ref.uri()).get()
print(f"Retrieved model: system_message={retrieved_model.system_message[:60]}...")


# =============================================================================
# 2. weave.StringPrompt / weave.MessagesPrompt - Prompt management
# =============================================================================
print("\n" + "=" * 60)
print("2. weave.StringPrompt / weave.MessagesPrompt - Prompt management")
print("=" * 60)

# StringPrompt - single string prompt
system_prompt = weave.StringPrompt("You speak like a friendly pirate.")
prompt_ref = weave.publish(system_prompt, name="pirate_prompt")
save_asset_ref("prompts", "pirate_prompt", prompt_ref.uri())

messages = [
    {"role": "system", "content": system_prompt.format()},
    {"role": "user", "content": "What is machine learning?"},
]
response = chat_completion(messages, max_tokens=80)
print(f"StringPrompt response: {response[:100]}...")

# MessagesPrompt - chat-message prompt template
messages_prompt = weave.MessagesPrompt(
    [
        {"role": "system", "content": "You are a helpful assistant specializing in {domain}."},
        {"role": "user", "content": "{question}"},
    ]
)
messages_prompt_ref = weave.publish(messages_prompt, name="domain_expert_prompt")
save_asset_ref("prompts", "domain_expert_prompt", messages_prompt_ref.uri())

formatted = messages_prompt.format(
    domain="machine learning",
    question="What is overfitting?",
)
response = chat_completion(formatted, max_tokens=80)
print(f"MessagesPrompt response: {response[:100]}...")

loaded_prompt = weave.ref(messages_prompt_ref.uri()).get()
print(f"Loaded prompt messages: {loaded_prompt.format(domain='AI', question='What is Weave?')}")


# =============================================================================
# 3. weave.Dataset - Create and retrieve datasets
# =============================================================================
print("\n" + "=" * 60)
print("3. weave.Dataset - Create and retrieve datasets")
print("=" * 60)

grammar_dataset = Dataset(
    name="grammar_benchmark",
    rows=[
        {"id": "0", "sentence": "He no likes ice cream.", "correction": "He doesn't like ice cream."},
        {"id": "1", "sentence": "She goed to the store.", "correction": "She went to the store."},
        {"id": "2", "sentence": "They was playing outside.", "correction": "They were playing outside."},
        {"id": "3", "sentence": "I has a big dog.", "correction": "I have a big dog."},
        {"id": "4", "sentence": "We runned very fast.", "correction": "We ran very fast."},
    ],
)
dataset_ref = weave.publish(grammar_dataset)
save_asset_ref("datasets", "grammar_benchmark", dataset_ref.uri())
print(f"Published dataset: {len(grammar_dataset.rows)} rows")


@weave.op()
def dummy_model(sentence: str) -> str:
    return sentence.upper()


_, call_1 = dummy_model.call("hello world")
_, call_2 = dummy_model.call("weave is great")
calls_dataset = Dataset.from_calls([call_1, call_2])
calls_dataset_ref = weave.publish(calls_dataset, name="calls_dataset")
save_asset_ref("datasets", "calls_dataset", calls_dataset_ref.uri())
print(f"Dataset from calls: {len(calls_dataset.rows)} rows")

retrieved_dataset = weave.ref(dataset_ref.uri()).get()
print(f"Retrieved dataset: {len(retrieved_dataset.rows)} rows")
print(f"First row: {dict(retrieved_dataset.rows[0])}")


# =============================================================================
# 4. Basic Scorer - Function-based
# =============================================================================
print("\n" + "=" * 60)
print("4. Basic Scorer - Function-based")
print("=" * 60)


@weave.op()
def exact_match_scorer(expected: str, output: dict) -> dict:
    """Check exact match, ignoring case."""
    generated = output.get("answer", "") if isinstance(output, dict) else str(output)
    return {"exact_match": expected.lower().strip() in generated.lower().strip()}


@weave.op()
def contains_answer_scorer(expected: str, output: dict) -> dict:
    """Check whether the expected answer is contained in the output."""
    generated = output.get("answer", "") if isinstance(output, dict) else str(output)
    return {"contains_answer": expected.lower() in generated.lower()}


result = exact_match_scorer("Paris", {"answer": "The capital is Paris."})
print(f"Exact match: {result}")
result = contains_answer_scorer("Paris", {"answer": "The capital is Paris."})
print(f"Contains answer: {result}")


# =============================================================================
# 5. Class-based Scorer
# =============================================================================
print("\n" + "=" * 60)
print("5. Class-based Scorer")
print("=" * 60)


class LengthScorer(Scorer):
    """Check response length."""

    min_length: int = 10
    max_length: int = 500

    @weave.op
    def score(self, output: str) -> dict:
        text = output.get("answer", "") if isinstance(output, dict) else str(output)
        length = len(text)
        return {
            "length": length,
            "is_appropriate": self.min_length <= length <= self.max_length,
        }


length_scorer = LengthScorer(min_length=20, max_length=200)
length_ref = weave.publish(length_scorer, name="length_scorer")
save_asset_ref("scorers", "length_scorer", length_ref.uri())

result = length_scorer.score({"answer": "This is a test response with some content."})
print(f"Length score: {result}")


# =============================================================================
# 6. LLM as a Judge
# =============================================================================
print("\n" + "=" * 60)
print("6. LLM as a Judge")
print("=" * 60)


class LLMJudgeScorer(Scorer):
    """Use an LLM as an evaluator."""

    criteria: str = "helpfulness"

    @weave.op
    def score(self, question: str, output: dict) -> dict:
        text = output.get("answer", "") if isinstance(output, dict) else str(output)

        messages = [
            {
                "role": "system",
                "content": f"""Evaluate the response based on {self.criteria}.
Return JSON: {{"score": 1-5, "reasoning": "brief explanation"}}""",
            },
            {"role": "user", "content": f"Question: {question}\nResponse: {text}"},
        ]

        try:
            result = json.loads(chat_completion(messages, temperature=0))
        except Exception:
            result = {"score": 0, "reasoning": "Parse error"}

        return result


judge = LLMJudgeScorer(criteria="accuracy and clarity")
judge_ref = weave.publish(judge, name="llm_judge_scorer")
save_asset_ref("scorers", "llm_judge_scorer", judge_ref.uri())

result = judge.score("What is 2+2?", {"answer": "The answer is 4."})
print(f"LLM Judge score: {result}")


# =============================================================================
# 7. Built-in Scorers - Weave-provided scorers
# =============================================================================
print("\n" + "=" * 60)
print("7. Built-in Scorers - Weave-provided scorers")
print("=" * 60)

print("""
Built-in Scorers provided by Weave:
==================================

Installation: pip install weave[scorers]

1. ValidJSONScorer - JSON format validation
2. ValidXMLScorer - XML format validation
3. HallucinationFreeScorer - Hallucination detection
4. SummarizationScorer - Summarization quality evaluation
5. OpenAIModerationScorer - Content moderation
6. EmbeddingSimilarityScorer - Embedding similarity
7. PydanticScorer - Pydantic schema validation
8. ContextEntityRecallScorer (RAGAS) - Entity recall
9. ContextRelevancyScorer (RAGAS) - Context relevancy

LLM-based scorers integrate with litellm, so you can specify
models from many providers with model_id.
""")


# =============================================================================
# 8. Using Built-in Scorers - Example
# =============================================================================
print("\n" + "=" * 60)
print("8. Using Built-in Scorers - Example")
print("=" * 60)

try:
    from weave.scorers import HallucinationFreeScorer, ValidJSONScorer

    json_scorer = ValidJSONScorer()

    @weave.op()
    def generate_json() -> str:
        return '{"name": "test", "value": 42}'

    output, call = generate_json.call()
    result = asyncio.run(call.apply_scorer(json_scorer))
    print(f"ValidJSONScorer result: {result.result}")

    print("\nHallucinationFreeScorer:")
    print("  from weave.scorers import HallucinationFreeScorer")
    print("  scorer = HallucinationFreeScorer(model_id='openai/gpt-4o')")
    print("  # Requires a context column when used in Evaluation")

except ImportError:
    print("Built-in scorers are not installed. Run: pip install weave[scorers]")


print("\n" + "=" * 60)
print("Assets and Scorers Demo Complete!")
print("=" * 60)
print(f"\nRegistered asset references: {ASSETS_FILE}")
print("""
Summary:
- Create and publish weave.Model / Prompt / Dataset assets
- Define function-based and class-based Scorers
- Review the types and usage of Built-in Scorers

Check in Weave UI:
- Use Models / Prompts / Datasets to inspect published assets
- Use Scorers to inspect the evaluation functions you created
""")

File: weave_introduction_handson_typescript/jp/2_1_assets.ts
/**
 * 2_1: Assets and Scorers - アセット管理と評価関数
 *
 * TypeScript 対応: 一部対応
 * - Dataset / StringPrompt / MessagesPrompt / publish は TypeScript SDK で扱います。
 * - weave.Model は現状 Python SDK の API です。TypeScript SDK には public export されていません。
 * - Python の weave.Model / weave.Scorer 継承は、TypeScript では設定 object と function-based scorer に置き換えます。
 * - Built-in scorers は公式ドキュメント上、Python SDK のみです。
 */

import * as weave from "weave";

import {
  chatCompletion,
  getModelName,
  getOpenAIClient,
  initWeave,
  printSection,
} from "../src/config.js";

type GrammarRow = {
  id: string;
  sentence: string;
  correction: string;
};

type GrammarOutput = {
  corrected: string;
};

const client = await initWeave();

// =============================================================================
// SECTION 1: Model Config + Traced Function
// =============================================================================
printSection("1. Model Config + traced function - LLM ラッパーのバージョン管理");
console.log(`
weave.Model は現状 Python SDK の API です。
TypeScript SDK には public export されていないため、この教材では
model config object と traced function の組み合わせで表現します。
`);

const grammarCorrectorConfigV2 = {
  objectType: "grammar_corrector_config",
  systemMessage: "You are an expert grammar checker. Return only the corrected sentence, no explanation.",
  modelName: getModelName(),
  temperature: 0,
};

const correctGrammar = weave.op(async function correctGrammar(sentence: string): Promise<GrammarOutput> {
  const response = await getOpenAIClient().chat.completions.create({
    model: grammarCorrectorConfigV2.modelName,
    messages: [
      { role: "system", content: grammarCorrectorConfigV2.systemMessage },
      { role: "user", content: sentence },
    ],
    temperature: grammarCorrectorConfigV2.temperature,
  });
  return { corrected: response.choices[0]?.message?.content?.trim() ?? "" };
});

const modelResult = await correctGrammar("I has a big dog.");
console.log(`Corrected: ${JSON.stringify(modelResult)}`);
console.log(`
注意:
- この config object はローカルの設定値です。
- TypeScript SDK には weave.Model がないため、これを Weave の Model asset として publish していません。
- correctGrammar(...) の実行内容は Weave の Traces に記録されます。
`);

// =============================================================================
// SECTION 2: Prompt Assets
// =============================================================================
printSection("2. StringPrompt / MessagesPrompt - プロンプト管理");

const systemPrompt = new weave.StringPrompt({
  content: "You speak like a friendly pirate.",
});
const promptRef = await client.publish(systemPrompt, "pirate_prompt_ts");
console.log(`Published StringPrompt: ${promptRef.uri()}`);

const pirateResponse = await chatCompletion(
  [
    { role: "system", content: systemPrompt.format() },
    { role: "user", content: "What is machine learning?" },
  ],
  { maxTokens: 80 }
);
console.log(`StringPrompt response: ${pirateResponse.slice(0, 100)}...`);

const messagesPrompt = new weave.MessagesPrompt({
  messages: [
    { role: "system", content: "You are a helpful assistant specializing in {domain}." },
    { role: "user", content: "{question}" },
  ],
});
const messagesPromptRef = await client.publish(messagesPrompt, "domain_expert_prompt_ts");
console.log(`Published MessagesPrompt: ${messagesPromptRef.uri()}`);

const formattedMessages = messagesPrompt.format({
  domain: "machine learning",
  question: "What is overfitting?",
});
const promptResponse = await chatCompletion(formattedMessages as any, { maxTokens: 80 });
console.log(`MessagesPrompt response: ${promptResponse.slice(0, 100)}...`);

// =============================================================================
// SECTION 3: Dataset Asset
// =============================================================================
printSection("3. Dataset - データセットの作成・保存");

const grammarDataset = new weave.Dataset<GrammarRow>({
  name: "grammar_benchmark_ts",
  rows: [
    { id: "0", sentence: "He no likes ice cream.", correction: "He doesn't like ice cream." },
    { id: "1", sentence: "She goed to the store.", correction: "She went to the store." },
    { id: "2", sentence: "They was playing outside.", correction: "They were playing outside." },
    { id: "3", sentence: "I has a big dog.", correction: "I have a big dog." },
    { id: "4", sentence: "We runned very fast.", correction: "We ran very fast." },
  ],
});
const datasetRef = await grammarDataset.save();
console.log(`Published dataset: ${datasetRef.uri()}`);
console.log(`Dataset rows: ${grammarDataset.length}`);
console.log(`First row: ${JSON.stringify(grammarDataset.getRow(0))}`);

// =============================================================================
// SECTION 4: Function-based Scorers
// =============================================================================
printSection("4. Function-based Scorers - 関数ベース");

const exactMatchScorer = weave.op(
  function exactMatchScorer({
    modelOutput,
    datasetRow,
  }: {
    modelOutput: GrammarOutput;
    datasetRow: GrammarRow;
  }) {
    return {
      exact_match: modelOutput.corrected.toLowerCase().trim() === datasetRow.correction.toLowerCase().trim(),
    };
  },
  { name: "exact_match_scorer_ts" }
);

const containsAnswerScorer = weave.op(
  function containsAnswerScorer({
    modelOutput,
    datasetRow,
  }: {
    modelOutput: GrammarOutput;
    datasetRow: GrammarRow;
  }) {
    return {
      contains_answer: modelOutput.corrected.toLowerCase().includes(datasetRow.correction.toLowerCase()),
    };
  },
  { name: "contains_answer_scorer_ts" }
);

console.log(
  `Exact match sample: ${JSON.stringify(
    await exactMatchScorer({
      modelOutput: { corrected: "I have a big dog." },
      datasetRow: grammarDataset.getRow(3),
    })
  )}`
);
console.log(
  `Contains sample: ${JSON.stringify(
    await containsAnswerScorer({
      modelOutput: { corrected: "I have a big dog." },
      datasetRow: grammarDataset.getRow(3),
    })
  )}`
);

// =============================================================================
// SECTION 5: Built-in Scorers Notes
// =============================================================================
printSection("5. Built-in Scorers - TypeScript 版での扱い");
console.log(`
Weave の Built-in / local scorers は Python SDK のみです。
TypeScript では上の exactMatchScorer のように function-based scorer を作り、
Evaluation に渡します。
`);

// =============================================================================
// COMPLETE
// =============================================================================
printSection("Assets and Scorers Demo Complete!");
console.log(`
まとめ:
- Dataset / Prompt は TypeScript SDK で保存可能
- weave.Model は現状 Python SDK のみ。TypeScript では Model asset として保存せず、traced function で実行を記録
- Python の weave.Scorer class / built-in scorers は function-based scorer に置換
`);
